rpc: Strict type checking for RPC boolean parameters

maflcko opened the pull request:
They are confusing, because:

* `getBool` is different from `get_bool`, but not documented as being different
* `isTrue` and `isFalse` return `false` for types other than `VBOOL`

This is a source of bugs, such as https://github.com/bitcoin/bitcoin/pull/26147#discussion_r983741878

MarcoFalke commented:
There are many more bugs, but to mention another one. This diff passes on master, but fails on the changes here:

```diff
diff --git a/test/functional/rpc_net.py b/test/functional/rpc_net.py
index ad8ba06824..f59bd81e20 100755
--- a/test/functional/rpc_net.py
+++ b/test/functional/rpc_net.py
@@ -190,7 +190,7 @@ class NetTest(BitcoinTestFramework):
 
         # Add an IPv6 address to the address manager.
         ipv6_addr = "1233:3432:2434:2343:3234:2345:6546:4534"
-        self.nodes[0].addpeeraddress(address=ipv6_addr, port=8333)
+        self.nodes[0].addpeeraddress(address=ipv6_addr, port=8333, tried={"foo":3})
 
         # Add 10,000 IPv4 addresses to the address manager. Due to the way bucket
         # and bucket positions are calculated, some of these addresses will collide.

luke-jr commented:
Suggest splitting up any fixes from the API change at least, so the fixes can be backported.

MarcoFalke commented:
There is no API change, but I've split the commit into three (bugfix / refactor)

furszy commented on src/util/system.cpp:
```diff
@@ -492,7 +492,7 @@ std::vector<std::string> ArgsManager::GetArgs(const std::string& strArg) const
     std::vector<std::string> result;
     for (const util::SettingsValue& value : GetSettingsList(strArg)) {
-        result.push_back(value.isFalse() ? "0" : value.isTrue() ? "1" : value.get_str());
+        result.push_back(value.isBool() ? (value.get_bool() ? "1" : "0") : value.get_str());
```
not for this PR, but would be good to place the "1" and "0" strings in constants, so we don't hardcode them when they are needed (here, in `SettingToString`, `SettingToInt`, and inside the univalue implementation).

MarcoFalke replied:
I presume `SettingsValue` is supposed to be independent of `UniValue` implementation details. Otherwise we could just return the UniValue internal string representation?

furszy replied:
ok yeah, agree.
leaving out the `UniValue` implementation, at least all settings methods could share the same constant (`SettingToString`, `ArgsManager::SoftSetBoolArg`, `ArgsManager::GetArgs`).

but.. just a small thing anyway.

MarcoFalke replied:
Maybe a `inline std::string BoolToSetting(bool val) { return val ? "1" : "0"; }` helper?

furszy replied:
sounds good 👍🏼

furszy reviewed (approved):
ACK 4444376c

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -729,9 +729,7 @@ static RPCHelpMan setban()
-        bool absolute = false;
-        if (request.params[3].isTrue())
-            absolute = true;
+        const bool absolute{request.params[3].isNull() ? false : request.params[3].get_bool()};
```
why?

furszy replied:
it's just a tiny nit, same expression. No need to apply it if you don't like it.

MarcoFalke replied:
I think it is worse, because it will make it impossible to change the default value without also changing the code logic.

luke-jr replied:
Agree with Marco, the current version is better

ryanofsky commented:
I can definitely understand why the `getBool` alias for `isTrue` is a footgun and should be removed. I also think the changes in this PR replacing `isTrue` with `get_bool` are appropriate in places where values are type checked, and you want to throw an exception when an unexpected type is encountered.

But in the places where you are replacing `isFalse` and `isTrue` with things like `isBool && get_bool` I think it makes code less clear. `getBool` really seems like a footgun. But `isFalse` `isTrue` don't seem like problems any more than `isNull` is.  They just seem like things that are appropriate to use some places, not appropriate other places.

ryanofsky commented:
Also agree with luke it would be nice to put the bugfix in a separate minimal PR so it easier to backport and easier to see what the actual bug is.

ryanofsky commented:
> The commit starts with "bugfix:", so it should be possible to find it.

I'm sure it's possible to figure it out, but it would be easier if something said what the bug was: steps to reproduce, expected, behavior, actual behavior

MarcoFalke commented:
> steps to reproduce, expected behavior, actual behavior

This is also there, see https://github.com/bitcoin/bitcoin/pull/26213#issuecomment-1263530762

MarcoFalke commented:
Rebased. Should be trivial to re-review.

aureleoules reviewed (approved):
ACK bbbb41ecd586db4ed51b8bbb217af26e04ff0315

MarcoFalke commented:
@ryanofsky Did you want to re-NACK this?

ryanofsky reviewed (approved):
I think there are still some issues with this PR, and at a minimum the commit descriptions should be updated to more clearly describe what is changing (suggestions in review comments below). But the code looks correct so I will add my code review ACK bbbb41ecd586db4ed51b8bbb217af26e04ff0315. I'd much more happily ACK this if it:

1. Had better commit descriptions that made it obvious how behavior was changing or not changing
2. Moved the third commit changing the UniValue API (bbbb41ecd586db4ed51b8bbb217af26e04ff0315) into a separate PR from the first commit changing RPC behavior (fa493c77965778c54b1206b3f8ab11c9b4d3beb3). The small second cleanup commit (fa671a15d3ab5f2466ee1066f6388ec67bc313ac) could go in either of these two PRs. But trying to do RPC bugfixes and UniValue API changes in the same PR IMO makes the RPC bugfixes more confusing to understand, and the UniValue API change harder to discuss and agree about.

Commit:
refactor: Replace isTrue with get_bool

This makes the code more robust, see previous commit.

In general replacing isTrue with get_bool is not equivalent because
get_bool can throw exceptions, but in this case, exceptions won't happen
because of RPCTypeCheck() and isNull() checks in the preceding code.

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -942,7 +940,7 @@ static RPCHelpMan addpeeraddress()
     const std::string& addr_string{request.params[0].get_str()};
     const auto port{request.params[1].getInt<uint16_t>()};
-    const bool tried{request.params[2].isTrue()};
+    const bool tried{request.params[2].isNull() ? false : request.params[2].get_bool()};
```
Nice. I've stolen the description from you to add a release note. I've added a test. I've dropped the third commit for now.

MarcoFalke commented on src/univalue/include/univalue.h:
```diff
@@ -66,16 +66,13 @@ class UniValue {
 
     bool isNull() const { return (typ == VNULL); }
-    bool isTrue() const { return (typ == VBOOL) && (val == "1"); }
-    bool isFalse() const { return (typ == VBOOL) && (val != "1"); }
```
Removed commit for now

MarcoFalke commented:
Force pushed to add a test and docs, thanks!

Commit:
bugfix: Strict type checking for RPC boolean parameters

ryanofsky reviewed (commented):
Code review ACK fa0153e609caf61a59efb0779e754861edc1684d

Looks great, thanks for taking various suggestions!

furszy reviewed (approved):
Code ACK fa0153e6
